### PR TITLE
[WIP] Update point goal while doing point nav for mobile gaze design

### PR DESCRIPTION
**Pull request by jimmytyyang**

### CODE REVIEW

**Comment by jimmytyyang** [comment]
@tusharsangam @KavitShah1998 This is a WIP PR for changing the point goal location while the robot is using a nav skill. I need your help to complete the function that computes the distance given the arm depth and rgb images. Thanks a lot!

**Comment by KavitShah1998** [comment]
For 3D object Detection, please refer to this branch - `experiment/3d_object_detection_with_spotcamera`

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 1]
```diff
@@ -587,6 +602,15 @@ def get_arm_joint_positions(self, as_array=True):
...
         return arm_joint_positions.tolist()
 
+    def get_new_goal_given_obj_img(self, depth, bbox):
+        """Estimate the goal location based on the depth and bbox of the target object
+        :param as_array
+        :return the x, y location in the world frame
+        """
+        x = 0
+        y = 0
+        return np.array([x, y])
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/spot.py:612]: Please see the `main()` method from HEAD of https://github.com/facebookresearch/spot-sim2real/tree/experiment/3d_object_detection_with_spotcamera for this implementation @tusharsangam

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 2]
```diff
@@ -207,7 +208,20 @@ def get_success(self, observations):
         return succ
 
     def get_observations(self):
-        return self.get_nav_observation(self.goal_xy, self.goal_heading)
+        # Update the point goal given depth and rgb images
+        if self._enable_nav_goal_change:
+            arm_depth, arm_depth_bbox = self.get_gripper_images()
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:213]: The depth returned by this method is "scaled". So we cannot use this method as is.

We need to add a new publisher for publishing "unscaled" / "raw" depth as obtained from BD (Check `SpotLocalRawImagesPublisher` in https://github.com/facebookresearch/spot-sim2real/tree/experiment/3d_object_detection_with_spotcamera which adds that additional publisher)

We need unscaled depth to get true reading of depth as a pixel coordinate (u,v)

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 3]
```diff
@@ -207,7 +208,20 @@ def get_success(self, observations):
...
+            # Update the object x, y here
+            obj_xy = self.spot.get_new_goal_given_obj_img(arm_depth, arm_depth_bbox)
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:215]: `arm_depth_bbox` should be `hand_rgb_bbox` IMO

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 4]
```diff
@@ -463,6 +463,24 @@ def set_base_velocity(
 
         return cmd_id
 
+    def get_global_from_local_based_on_input_T(
+        self, x_pos, y_pos, yaw, curr_x, curr_y, curr_yaw
+    ):
+        """Transform the point given the x,y,yaw location"""
+        coors = np.array([x_pos, y_pos, 1.0])
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/spot.py:470]:
```suggestion
        global_coords = np.array([x_pos, y_pos, 1.0])
```

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 5]
```diff
@@ -463,6 +463,24 @@ def set_base_velocity(
...
+        local_T_global = self._get_local_T_global(curr_x, curr_y, curr_yaw)
+        x, y, w = local_T_global.dot(coors)
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/spot.py:472]: Should it be 

```suggestion
        x, y, z = local_T_global.dot(coors)
```
??

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 6]
```diff
@@ -463,6 +463,24 @@ def set_base_velocity(
...
+        global_x_pos, global_y_pos = x / w, y / w
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/spot.py:473]: Why are we scaling this with "z" ?

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 7]
```diff
@@ -207,7 +208,20 @@ def get_success(self, observations):
...
+            # Update the obj_xy to consider the base offset
+            obj_nav_offset_x = -1.5
+            goal_x, goal_y, _ = self.spot.get_global_from_local_based_on_input_T(
+                obj_nav_offset_x, 0, 0, obj_xy[0], obj_xy[1], 0
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:219]: If this offset is in global frame, then adjustment along only x axis may not solve the issue. We may need a more generalizable approximation function.

@jimmytyyang  as we discussed

**Comment by tusharsangam** [comment]
def detect_3d_point_from_image_coordinate(Z:float, rgb_uv, camera_intrinsics):
    u, v = rgb_uv
    focal_x, focal_y = camera_intrinsics.focal_length.x, camera_intrinsics.focal_length.y
    principal_point_x, principal_point_y = camera_intrinsics.principal_point.x, camera_intrinsics.principal_point.y
    X = (rgb_uv[0] - principal_point_x)*Z / focal_x
    Y = (rgb_uv[1] - principal_point_y)*Z/ focal_y
    return (X,Y,Z)
We convert X.Y,Z coordinate back to dock coordinates

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 7 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:219]: @tusharsangam you might want to rewrite something here :)

**Pull request closed without merging**